File: 2026-03-March-GFG-POTD/March 16 - K Sum Paths/Solution.java
/*
class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}
*/

class Solution {
    public int countAllPaths(Node root, int k) {
        HashMap<Long, Integer> map = new HashMap<>();
        map.put(0L, 1); // base case
        
        return dfs(root, 0, k, map);
    }
    
    private int dfs(Node node, long currSum, int k, HashMap<Long, Integer> map) {
        if (node == null) return 0;
        
        currSum += node.data;
        
        // Check if there exists a prefix sum we can subtract
        int count = map.getOrDefault(currSum - k, 0);
        
        // Add current sum to map
        map.put(currSum, map.getOrDefault(currSum, 0) + 1);
        
        // Traverse left and right
        count += dfs(node.left, currSum, k, map);
        count += dfs(node.right, currSum, k, map);
        
        // Backtrack (important!)
        map.put(currSum, map.get(currSum) - 1);
        
        return count;
    }
}

File: 2026-03-March-GFG-POTD/March 16 - K Sum Paths/Solution.cpp
#include <unordered_map>
using namespace std;

/*
struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int x) {
        data = x;
        left = right = NULL;
    }
};
*/

class Solution {
public:
    int countAllPaths(Node* root, int k) {
        unordered_map<long long, int> mp;
        mp[0] = 1; // base case
        return dfs(root, 0, k, mp);
    }

private:
    int dfs(Node* node, long long currSum, int k, unordered_map<long long, int>& mp) {
        if (!node) return 0;

        currSum += node->data;

        int count = mp[currSum - k];

        mp[currSum]++;

        count += dfs(node->left, currSum, k, mp);
        count += dfs(node->right, currSum, k, mp);

        mp[currSum]--; // backtrack

        return count;
    }
};
